src/main/java/net/snowflake/ingest/streaming/internal/DataValidationUtil.java

```diff
@@ -103,21 +111,38 @@ private static BigDecimal[] makePower10Table() {
    * see {@link DataValidationUtil#isAllowedSemiStructuredType}.
    *
    * @param input Object to validate
-   * @return JSON tree representing the input
+   * @return Minified JSON string
    */
-  private static JsonNode validateAndParseSemiStructuredAsJsonTree(
+  private static String validateAndParseSemiStructured(
       String columnName, Object input, String snowflakeType, final long insertRowIndex) {
     if (input instanceof String) {
-      String stringInput = (String) input;
+      final String stringInput = (String) input;
       verifyValidUtf8(stringInput, columnName, snowflakeType, insertRowIndex);
-      try {
-        return objectMapper.readTree(stringInput);
-      } catch (JsonProcessingException e) {
+      final StringBuilderWriter resultWriter = new StringBuilderWriter(stringInput.length());
+      try (final JsonParser parser = factory.createParser(stringInput);
+          final JsonGenerator generator = factory.createGenerator(resultWriter)) {
+        while (parser.nextToken() != null) {
+          final JsonToken token = parser.currentToken();
+          if (token.isNumeric()) {
+            // If the current token is a number, we cannot just copy the current event because it
+            // would write token the token from double (or big decimal), whose scientific notation
+            // may have been altered during deserialization. We want to preserve the scientific
+            // notation from the user input, so we write the current numer as text.
+            generator.writeNumber(parser.getText());
+          } else {
+            generator.copyCurrentEvent(parser);
+          }
+        }
+      } catch (JsonParseException e) {
         throw valueFormatNotAllowedException(
```

Comment:
we're losing all information about the original exception (e) in this throw. can `e` be added as an inner exception ? Or log it out before throwing?

Reply:
> This is on purpose. This exception is thrown when customers pass a value, which cannot be ingested into the target column type, e.g. because the value does not match the column type, or the value is too large for the column.

Wait, why this is on purpose? It would be good to throw the original exception, then customer can look at it in the error response and fix the input?

Reply:
> and we must not put customer data into logs

Agree we should not put it in the log, what I thought is to include the original exception as part of the `InsertValidationResponse` so customer can audit the error and fix it if needed.

Reply:
We can do that, but as a separate PR. My PR is not introducing any changes in how our validation exceptions look like.